src/uct/ib/mlx5/dv/ib_mlx5_dv.c

```diff
@@ -26,6 +28,237 @@ ucs_status_t uct_ib_mlx5dv_init_obj(uct_ib_mlx5dv_t *obj, uint64_t type)
+                                      struct ibv_ah_attr *ah_attr)
+{
+#if HAVE_DEVX
+    uint32_t in_2rtr[UCT_IB_MLX5DV_ST_SZ_DW(init2rtr_qp_in)] = {};
+    uint32_t out_2rtr[UCT_IB_MLX5DV_ST_SZ_DW(init2rtr_qp_out)] = {};
+    uint32_t in_2rts[UCT_IB_MLX5DV_ST_SZ_DW(rtr2rts_qp_in)] = {};
+    uint32_t out_2rts[UCT_IB_MLX5DV_ST_SZ_DW(rtr2rts_qp_out)] = {};
+    void *qpc;
+    int ret;
+
+    ucs_assert(qp->type == UCT_IB_MLX5_QP_TYPE_DEVX);
+    UCT_IB_MLX5DV_SET(init2rtr_qp_in, in_2rtr, opcode, UCT_IB_MLX5_CMD_OP_INIT2RTR_QP);
+    UCT_IB_MLX5DV_SET(init2rtr_qp_in, in_2rtr, qpn, qp->qp_num);
+    UCT_IB_MLX5DV_SET(init2rtr_qp_in, in_2rtr, opt_param_mask, 14);
+
+    qpc = UCT_IB_MLX5DV_ADDR_OF(init2rtr_qp_in, in_2rtr, qpc);
+    UCT_IB_MLX5DV_SET(qpc, qpc, mtu, 5);
+    UCT_IB_MLX5DV_SET(qpc, qpc, log_msg_max, 30);
+    UCT_IB_MLX5DV_SET(qpc, qpc, remote_qpn, dest_qp_num);
+    UCT_IB_MLX5DV_SET(qpc, qpc, primary_address_path.rlid, ah_attr->dlid);
+    UCT_IB_MLX5DV_SET(qpc, qpc, primary_address_path.grh, ah_attr->is_global);
+    memcpy(UCT_IB_MLX5DV_ADDR_OF(qpc, qpc, primary_address_path.rgid_rip),
+            &ah_attr->grh.dgid,
+            UCT_IB_MLX5DV_FLD_SZ_BYTES(qpc, primary_address_path.rgid_rip));
+    UCT_IB_MLX5DV_SET(qpc, qpc, primary_address_path.vhca_port_num, ah_attr->port_num);
+    UCT_IB_MLX5DV_SET(qpc, qpc, log_ack_req_freq, 8);
+    UCT_IB_MLX5DV_SET(qpc, qpc, log_rra_max, 2);
+    UCT_IB_MLX5DV_SET(qpc, qpc, atomic_mode, 5);
+    UCT_IB_MLX5DV_SET(qpc, qpc, rre, 1);
+    UCT_IB_MLX5DV_SET(qpc, qpc, rwe, 1);
+    UCT_IB_MLX5DV_SET(qpc, qpc, rae, 1);
+    UCT_IB_MLX5DV_SET(qpc, qpc, min_rnr_nak, 13);
+
+    ret = mlx5dv_devx_obj_modify(qp->devx.obj, in_2rtr, sizeof(in_2rtr),
+                                 out_2rtr, sizeof(out_2rtr));
+    if (ret) {
+        ucs_error("Failed to modify QP to RTR %m");
+        return UCS_ERR_IO_ERROR;
+    }
+
+    UCT_IB_MLX5DV_SET(rtr2rts_qp_in, in_2rts, opcode, UCT_IB_MLX5_CMD_OP_RTR2RTS_QP);
+    UCT_IB_MLX5DV_SET(rtr2rts_qp_in, in_2rts, qpn, qp->qp_num);
+
+    qpc = UCT_IB_MLX5DV_ADDR_OF(rst2init_qp_in, in_2rts, qpc);
+    UCT_IB_MLX5DV_SET(qpc, qpc, log_ack_req_freq, 8);
+    UCT_IB_MLX5DV_SET(qpc, qpc, log_sra_max, 2);
+    UCT_IB_MLX5DV_SET(qpc, qpc, retry_count, 7);
+    UCT_IB_MLX5DV_SET(qpc, qpc, rnr_retry, 7);
+    UCT_IB_MLX5DV_SET(qpc, qpc, primary_address_path.ack_timeout, 18);
+
+    ret = mlx5dv_devx_obj_modify(qp->devx.obj, in_2rts, sizeof(in_2rts),
+                                 out_2rts, sizeof(out_2rts));
+    if (ret) {
+        ucs_error("Failed to modify QP to RTS %m");
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
+#else
+    return UCS_ERR_UNSUPPORTED;
+#endif
+}
+
+void uct_ib_mlx5dv_destroy_qp(uct_ib_mlx5_qp_t *qp)
+{
+#if HAVE_DEVX
+    mlx5dv_devx_obj_destroy(qp->devx.obj);
+    ucs_mpool_put_inline(qp->devx.dbrec);
+    if (qp->devx.wq_buf != NULL) {
+        mlx5dv_devx_umem_dereg(qp->devx.mem);
+        free(qp->devx.wq_buf);
```

Comment:
ucs_free
shouldn't it be outside the 'if'?
also, maybe mlx5dv_devx_umem_dereg() should be able to handle NULL (ignore it)?